### PR TITLE
Add .NET Core 3.1 build of engine to access APIs for loading .NET Core assemblies correctly

### LINKED ISSUE #710: .NET Core engine only works when located in same directory as test assembly
I am trying to run tests using NUnit.Engine.Api in Net Core, but I am getting this error:

Could not load file or assembly 'NUnitTestNetCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.

My code is:

            ITestEngine nunitEngine = TestEngineActivator.CreateInstance();

            TestPackage package = new TestPackage("Test DLL Path");
            package.AddSetting("WorkDirectory", @"Test DLL WorkDirectory PATH");

            var filterService = nunitEngine.Services.GetService<ITestFilterService>();
            ITestFilterBuilder builder = filterService.GetTestFilterBuilder();

            TestFilter emptyFilter = builder.GetFilter();
            TestEventListner testListener = new TestEventListner();

            using (ITestRunner runner = nunitEngine.GetRunner(package))
            {
                XmlNode result = runner.Run(testListener, emptyFilter);
            }

I am using:
.NET Core 3.0
NUnit.Engine 3.10.0
NUnit.Engine.Api 3.10.0

### PR DESCRIPTION
**Pull request by ChrisMaddock**
To sort #710 properly, we need to make use of the AssemblyLoadContext APIs, which are only available in .NET Core 2.1 and above.

In our 'full vision', this functionality would live in a .NET Core specific agent, and not require a full build of the engine. However, I don't want to delay the .NET Core Console further while we work on .NET Core agents - and I consider #710 a major enough bug that I don't want to release the .NET Core Console without a solution.

Accordingly, this PR adds a .NET Core 3.1 build of the engine. We use v3.1 specifically to make use of the `AssemblyDependencyResolver` class, which was first available in .NET Core 3. (Of which, 3.1 was the first LTS release). There are workarounds for earlier frameworks - however my priority now is getting the latest version of .NET Core working. If we end up with separate agents for older versions of .NET Core, it would be possible for someone to look at adding the workarounds into those - if there's sufficient demand.

I expect this build to be a temporary build, until we have .NET Core agents, at which point we can remove it and go back to just packaging .NET Standard 2.0. I can't think of any way that could be a breaking change, on the basis that any .NET Core 3.1 runners can consume the .NET standard package instead?

---

For ease of review, this PR just includes adding the .NET Core 3.1 build. The actual fix will be in a second PR.

### CODE REVIEW

File: src/NUnitEngine/nunit.engine.tests/Drivers/NUnit3FrameworkDriverTests.cs [hunk 1]
```diff
@@ -51,21 +51,6 @@ public void CreateDriver()
...
-        //    Assert.That(_controller.AssemblyPath, Is.EqualTo(MOCK_ASSEMBLY));
-        //    Assert.That(_controller.Settings, Is.SameAs(_settings));
-        //}
-
-        public void ConstructController_MissingFile_ThrowsArgumentInvalid()
-        {
-            var assemblyName = typeof(NUnit.Framework.TestAttribute).Assembly.GetName();
-            Assert.That(new NUnit3FrameworkDriver(AppDomain.CurrentDomain, assemblyName), Throws.ArgumentException);
-        }
-
```
Comment (ChrisMaddock) [inline src/NUnitEngine/nunit.engine.tests/Drivers/NUnit3FrameworkDriverTests.cs:68]: This is unrelated, but this test was missing its TestAttribute so not running. The test is no longer valid.

File: src/NUnitEngine/nunit.engine.tests/Services/ExtensionServiceTests.cs [hunk 2]
```diff
@@ -256,8 +258,8 @@ public static IEnumerable<TestCaseData> InvalidTargetFrameworkCombos()
 
         public static IEnumerable<TestCaseData> InvalidRunnerCombos()
         {
-#if NETCOREAPP2_1
-            Assembly netstandard = typeof(ExtensionService).Assembly;
+#if NETCOREAPP
+            Assembly netstandard = Assembly.LoadFile(Path.Combine(GetSiblingDirectory("netstandard2.0"), "nunit.engine.dll"));
```
Comment (ChrisMaddock) [inline src/NUnitEngine/nunit.engine.tests/Services/ExtensionServiceTests.cs:262]: These tests were using the executing engine as an example .NET Standard assembly - which of course caused problems when that assembly is build for .NET Core. I changed them to fetch the .NET Standard build specifically, in the same way the other tests in this class do.

Ideally, I think we need a set of test assemblies for this. (Or perhaps to use mock-assembly?)

File: src/NUnitConsole/nunit3-console/nunit3-console.csproj [hunk 3]
```diff
@@ -3,7 +3,7 @@
     <OutputType>Exe</OutputType>
     <RootNamespace>NUnit.ConsoleRunner</RootNamespace>
     <AssemblyName>nunit3-console</AssemblyName>
-    <TargetFrameworks>net20;netcoreapp2.1</TargetFrameworks>
```
Comment (ChrisMaddock) [inline src/NUnitConsole/nunit3-console/nunit3-console.csproj:6]: The console is as-yet unreleased, but will now target .NET Core 3.1 as a minimum. With the long-term aim of making it self-contained anyway, I'm not too worried about this.

**Comment by ChrisMaddock** [comment]
Just added the blocked label, as I'll need to rebase this over #779 once that's merged, and make the same change for the new .NET Core 3.1 build. That's only a minor change however, the rest is still ready for review. 🙂

**Comment by ChrisMaddock** [comment]
Reverting to draft until I can figure out what's causing the test failures seen in #781

**Comment by ChrisMaddock** [comment]
Ok - I'm happy this is the right approach now and everything's working - I'm just going to tidy up the commits, and it'll be ready for review. 🙂

**Comment by ChrisMaddock** [comment]
Ok - this is now ready for review. I've tried to break things up so it's easy to look at commit-by-commit.

As a reminder, this PR only includes adding the .NET Core 3.1 build itself - the fix will come in a later PR.

File: build.cake [hunk 4]
```diff
@@ -202,28 +201,28 @@ Task("Build")
...
                 .WithProperty("PublishDir", BIN_DIR + framework));
 
-        foreach(var framework in NETSTANDARD_FRAMEWORKS)
+        foreach(var framework in new [] { "netstandard1.6", "netstandard2.0" })
              MSBuild(ENGINE_API_CSPROJ, CreateMSBuildSettings("Publish")
                 .WithProperty("TargetFramework", framework)
                 .WithProperty("PublishDir", BIN_DIR + framework));
 
-        foreach(var framework in NETCORE_FRAMEWORKS)
+        foreach(var framework in new [] { "netcoreapp1.1", "netcoreapp2.1", "netcoreapp3.1" })
```
Comment (jnm2) [inline build.cake:214]: Could we give this list of frameworks a descriptive name since it repeats?

File: build.cake [hunk 5]
```diff
@@ -667,7 +692,7 @@ FilePath GetResultXmlPath(string testAssembly, string framework)
 {
     var assemblyName = System.IO.Path.GetFileNameWithoutExtension(testAssembly);
 
-    // Workaround for https://github.com/nunit/nunit-console/issues/501
+    // Required for Test suites running under NUnitLite
```
Comment (jnm2) [inline build.cake:695]:
```suggestion
    // Required for test suites running under NUnitLite
```

**Comment by ChrisMaddock** [comment]
Thanks Joseph. 😄 

> Did I miss the place where the .NET Core APIs begin to be used to load assemblies, or is that for a later PR?

Yeah, that bit is in https://github.com/nunit/nunit-console/pull/781. 🙂 It's depends on three other open PRs, so I'll rebase and tidy up once all those are merged. 

Looks like there's a little more battle required yet, to get .NET Core 1.1 running on Azure. Might have to come back to that later, will sort your feedback at the same time. Thanks!

File: build.cake (hunk 4 above)
Comment (ChrisMaddock) [reply build.cake:214]: So the lists of targets for each build are now different. The engine publishes for netstandard1.6, netstandard2.0 and netcoreapp3.1. The API dll for the two netstandard targets, and the tests for three net core targets. 

That was my original reason for getting rid of the variables - and I'm not too sure how we could name the three different lists helpfully - being explicit and inlining the lists was the best I could come up with. Any suggestions?

**Comment by ChrisMaddock** [comment]
So the old Azure image won't build .NET Core 3.1, and the new one won't build .NET Core 1.1. 😤

Guess this is a task for another day...

**Comment by jnm2** [comment]
Just a suggestion and I might be missing some context, but how sure are we that .NET Core 1.1 is worth building for? It's past end of life.

File: build.cake (hunk 4 above)
Comment (jnm2) [reply build.cake:214]: Oh, I misread. This seems fine!

**Comment by ChrisMaddock** [comment]
Yeah, I'm inclined to agree, see https://github.com/nunit/nunit-console/issues/725. It's too much of a timesink for the benefit, in my opinion. I just feel like that should be done separately and intentionally, rather than taken out because something else broke it - but it does pull in to question how much of my time it's worth investing in trying to get it back up and running again...

**Comment by ChrisMaddock** [comment]
It's a little frustrating, as it would be a breaking change for runner-maintainers...however I expect the number of actual end users is critically small.

**Comment by ChrisMaddock** [comment]
Finally got CI working again. 

Thanks Joseph for reviewing this. 🙂  As it's a significant change, I'll just leave the PR up a few more days before merging, to give anyone else on the team the chance to raise any concerns.

**Pull request merged**